test/common/stats/thread_local_store_test.cc

```diff
@@ -190,6 +190,182 @@ TEST_F(StatsThreadLocalStoreTest, BasicScope) {
+  store_->initializeThreading(main_thread_dispatcher_, tls_);
+
+  Histogram& h1 = store_->histogram("h1");
+  Histogram& h2 = store_->histogram("h2");
+  EXPECT_EQ("h1", h1.name());
+  EXPECT_EQ("h2", h2.name());
+
+  EXPECT_CALL(sink_, onHistogramComplete(Ref(h1), 1));
+  h1.recordValue(1);
+
+  EXPECT_CALL(sink_, onHistogramComplete(Ref(h2), 1));
+  h2.recordValue(1);
+
+  EXPECT_CALL(sink_, onHistogramComplete(Ref(h2), 2));
+  h2.recordValue(2);
+
+  std::shared_ptr<std::atomic<bool>> merge_called = std::make_shared<std::atomic<bool>>(false);
+  store_->mergeHistograms([merge_called]() -> void { *merge_called = true; });
+
+  EXPECT_TRUE(*merge_called);
+
+  std::list<HistogramSharedPtr> histogram_list = store_->histograms();
+
+  EXPECT_EQ(histogram_list.size(), 2);
+  std::string h1_summary = "P0: 1, P25: 1.025, P50: 1.05, P75: 1.075, P90: 1.09, P95: 1.095, P99: "
+                           "1.099, P99.9: 1.0999, P100: 1.1";
+  std::string h2_summary = "P0: 1, P25: 1.05, P50: 1.1, P75: 2.05, P90: 2.08, P95: 2.09, P99: "
+                           "2.098, P99.9: 2.0998, P100: 2.1";
+
+  for (const Stats::HistogramSharedPtr& histogram : histogram_list) {
+    EXPECT_EQ(histogram->cumulativeStatistics().summary(),
+              histogram->intervalStatistics().summary());
+    if (histogram->name().find("h1") != std::string::npos) {
+      EXPECT_EQ(histogram->cumulativeStatistics().summary(), h1_summary);
+    } else {
+      EXPECT_EQ(histogram->cumulativeStatistics().summary(), h2_summary);
+    }
+  }
+
+  store_->shutdownThreading();
+  tls_.shutdownThread();
+
+  // Includes overflow stat.
+  EXPECT_CALL(*this, free(_));
+}
+
+TEST_F(StatsThreadLocalStoreTest, MultipleMerges) {
+  InSequence s;
+  store_->initializeThreading(main_thread_dispatcher_, tls_);
+
+  Histogram& h1 = store_->histogram("h1");
+  Histogram& h2 = store_->histogram("h2");
+  EXPECT_EQ("h1", h1.name());
+  EXPECT_EQ("h2", h2.name());
+
+  // Test Merge with inserting one value in to histogram and validate that summary returns correct
+  // values.
+  EXPECT_CALL(sink_, onHistogramComplete(Ref(h1), 1));
+  h1.recordValue(1);
+
+  std::shared_ptr<std::atomic<bool>> merge_called = std::make_shared<std::atomic<bool>>(false);
+
+  store_->mergeHistograms([merge_called]() -> void { *merge_called = true; });
+
+  EXPECT_TRUE(*merge_called);
+
+  std::list<HistogramSharedPtr> histogram_list = store_->histograms();
+
+  EXPECT_EQ(histogram_list.size(), 2);
+
+  std::string summary_with_one_as_value =
+      "P0: 1, P25: 1.025, P50: 1.05, P75: 1.075, P90: 1.09, P95: 1.095, P99: "
+      "1.099, P99.9: 1.0999, P100: 1.1";
+  std::string summary_with_two_as_value = "P0: 1, P25: 2.025, P50: 2.05, P75: 2.075, P90: 2.09, "
+                                          "P95: 2.095, P99: 2.099, P99.9: 2.0999, P100: 2.1";
+  std::string not_used_summary =
+      "P0: 0, P25: 0, P50: 0, P75: 0, P90: 0, P95: 0, P99: 0, P99.9: 0, P100: 0";
+  std::string not_used_in_interval_summary =
```

Comment:
@postwait after we call `hist_free` of an existing histogram and get quantiles on it we get all 1's instead of all 0's as we get with newly allocated histogram. Is that expected?

Reply:
You can't call quantiles on a histogram you've freed, that would crash or quantile trash.  Use `hist_clear` to remove all samples from a histogram (but leave the bins in place).  If you call the quantile function on a histogram and get 1's it is an indicator that it has value(s) that exist in the [1.0,1.1) bin.

Reply:
Sorry. I meant hist_clear only. After I call hist_clear I am getting all 1’s.

Reply:
i am using hist_alloc(). If i call hist_free and hist_alloc, it returns 0's but after hist_clear return 1's
Sample code that reproduces the problem
  histograms_[0] = hist_alloc();
  hist_insert_intscale(histograms_[0], 1, 0, 1);
  hist_clear(histograms_[0]);

Reply:
Try --config=asan or running under valgrind.

Reply:
Try --config=asan or running under valgrind.

Reply:
Try --config=asan or running under valgrind.

Reply:
Try --config=asan or running under valgrind.

Reply:
Test and fix added upstream.  Nice catch.  Edge case on empty histograms for quantiles.

Your output double vector will be full of NaN's now if you ask about an empty histogram.

Reply:
@postwait thanks a lot. I updated the new version and it works correctly.

Reply:
@postwait looks like found another corner case issue or may be i am misunderstanding.

Consider the code

    histogram_t* main_thread_interval_hist = hist_alloc();
    histogram_t* per_thread_interval_hist = hist_alloc();
    hist_insert_intscale(per_thread_interval_hist, 1, 0, 1);
    double in[9] = {0, 0.25, 0.5, 0.75, 0.90, 0.95, 0.99, 0.999, 1};
    double out[9];
    hist_approx_quantile(main_thread_interval_hist, in, 9,out);
    histogram_t* hist_array[1];
    hist_array[0] = per_thread_interval_hist;
    hist_accumulate(main_thread_interval_hist, hist_array, ARRAY_SIZE(hist_array));
    hist_clear(per_thread_interval_hist);
    //hist_free(per_thread_interval_hist);
    //per_thread_interval_hist = hist_alloc();
    hist_array[0] = per_thread_interval_hist;
    hist_insert_intscale(per_thread_interval_hist, 2, 0, 1);
    //hist_clear(main_thread_interval_hist);
    //hist_free(main_thread_interval_hist);
    main_thread_interval_hist = hist_alloc();
    hist_accumulate(main_thread_interval_hist, hist_array, ARRAY_SIZE(hist_array));
    hist_approx_quantile(main_thread_interval_hist, in, 9,out);
    std::cout<<"after accumulations :"<<out[0]<<"==> \n";

    histogram_t* direct_hist = hist_alloc();
    hist_insert_intscale(direct_hist, 2, 0, 1);
    hist_approx_quantile(direct_hist, in, 9,out);
    std::cout<<"direct :"<<out[0]<<"==> \n";

I expect direct and after accumulations to give the same output. But it gives 1 in first bin (rest of the bins are same) for accumulation case and 2 in the direct case.
If I change the code to free and hist_alloc it works correctly in both the cases. Is this is a corner case issue with hist_clear?

Reply:
@postwait Please see the above comment